ethers/providers/jsonrpc.nim

```diff
@@ -165,6 +189,91 @@ method signMessage*(signer: JsonRpcSigner,
+             wantedConfirms = DEFAULT_CONFIRMATIONS,
+             timeoutInBlocks = RECEIPT_TIMEOUT_BLKS.some): # will error if tx not mined in x blocks
+            Future[TransactionReceipt]
+            {.async, upraises: [JsonRpcProviderError].} = # raises for clarity
+
+  var
+    receipt: ?TransactionReceipt
+    subscription: JsonRpcSubscription
+
+  let
+    provider = JsonRpcProvider(tx.provider)
+    retFut = newFuture[TransactionReceipt]("wait")
+
+  proc confirmations(receipt: TransactionReceipt, atBlkNum: UInt256): UInt256 =
+
+    var confirms = (atBlkNum - !receipt.blockNumber) + 1
+    if confirms <= 0: confirms = 1.u256
+    return confirms
+
+  proc newBlock(blk: Block) =
+    # has been mined, need to check # of confirmations thus far
+    let confirms = (!receipt).confirmations(blk.number)
+    if confirms >= wantedConfirms.u256:
+      # fire and forget
+      discard subscription.unsubscribe()
+      retFut.complete(!receipt)
+
+  let startBlock = await provider.getBlockNumber()
+
+  # loop until the tx is mined, or times out (in blocks) if timeout specified
+  while receipt.isNone:
+    receipt = await provider.getTransactionReceipt(tx.hash)
+    if receipt.isSome and (!receipt).blockNumber.isSome:
+      break
+
+    if timeoutInBlocks.isSome:
+      let currBlock = await provider.getBlockNumber()
+      let blocksPassed = (currBlock - startBlock) + 1
+      if blocksPassed >= (!timeoutInBlocks).u256:
+        raiseProviderError("Transaction was not mined in " &
+          $(!timeoutInBlocks) & " blocks")
+
+    await sleepAsync(RECEIPT_POLLING_INTERVAL.seconds)
+
+  # has been mined, need to check # of confirmations thus far
+  let confirms = (!receipt).confirmations(startBlock)
+  if confirms >= wantedConfirms.u256:
+    return !receipt
+
+  else:
+    let sub = await provider.subscribe(newBlock)
+    subscription = JsonRpcSubscription(sub)
+    return (await retFut)
+
+method wait*(tx: Future[TransactionResponse],
+             wantedConfirms = DEFAULT_CONFIRMATIONS,
+             timeoutInBlocks = RECEIPT_TIMEOUT_BLKS.some):
+            Future[TransactionReceipt] {.async.} =
+  ## Convenience method that allows wait to be chained to a sendTransaction
+  ## call, eg:
+  ## `await signer.sendTransaction(populated).wait(3)`
+
+  let txResp = await tx
+  return await txResp.wait(wantedConfirms, timeoutInBlocks)
+
+method wait*(tx: Future[?TransactionResponse],
+             wantedConfirms = DEFAULT_CONFIRMATIONS,
+             timeoutInBlocks = RECEIPT_TIMEOUT_BLKS.some):
+            Future[TransactionReceipt] {.async.} =
+  ## Convenience method that allows wait to be chained to a contract
+  ## transaction, eg:
+  ## `await token.connect(signer0)
+  ##          .mint(accounts[1], 100.u256)
+  ##          .wait(3)`
+
+  let txResp = await tx
+  if txResp.isNone:
+    raiseProviderError("Transaction hash required. Possibly was a call instead of a send?")
+
+  return await (!txResp).wait(wantedConfirms, timeoutInBlocks)
```

Comment:
```suggestion
  without txResp =? (await tx):
    raiseProviderError("Transaction hash required. Possibly was a call instead of a send?")

  return await txResp.wait(wantedConfirms, timeoutInBlocks)
```

Reply:
Much better, thanks.